cloudmarker/clouds/gcpcloud.py

```diff
@@ -71,213 +88,186 @@ def _get_resource(self, service_name, version='v1'):
-                        'record_type': 'compute'
-                    }
+        record_type_map = {
+            'compute': 'compute',
+        }
+        for i, raw_record in enumerate(iterator):
+            record = {
+                'raw': raw_record,
+                'ext': {
+                    'cloud_type': 'gcp',
+                    'record_type': gcp_record_type,
+                    'project_id': project.get('projectId'),
+                    'project_name': project.get('name'),
+                    'zone': zone,
+                    'key_file_path': self._key_file_path,
+                    'client_email': self._client_email
+                },
+                'com': {
+                    'cloud_type': 'gcp',
+                    'record_type': record_type_map.get(gcp_record_type)
                 }
-                yield record
+            }
 
-            if i + 1 == self._max_projects:
-                _log.info('Stopping fetch for %s due to _max_projects: %d',
-                          project, self._max_projects)
-                break
+            _log.info('Found %s #%d: %s; %s', gcp_record_type, i,
+                      raw_record.get('name'),
+                      util.outline_gcp_project(project_index, project, zone,
+                                               self._key_file_path))
+            yield record
+
+            if gcp_record_type == 'firewall':
+                yield from _get_normalized_firewall_rules(record,
+                                                          project_index,
+                                                          project,
+                                                          self._key_file_path)
 
     def done(self):
-        """Perform clean up tasks.
+        """Log a message that this plugin is done."""
+        _log.info('Done; key_file_path: %s; processes: %s; threads: %s',
+                  self._key_file_path, self._processes, self._threads)
 
-        Currently, this method does nothing because there are no clean
-        up tasks associated with the :class:`GCPCloud` plugin. This
-        may change in future.
-        """
 
+def _get_resource_iterator(resource, key, **list_kwargs):
+    """Generate records for specific record types.
+
+    Arguments:
+        resource (Resource): GCP resource object.
+        key (str): The key that we need to look up in the GCP
+            response JSON to find the list of resources.
+        list_kwargs (dict): Keyword arguments for
+            ``resource.list()`` call.
+
+    Yields:
+        dict: A GCP configuration record.
 
-def _get_normalized_firewall_rules(firewall_record):
+    """
+    request = resource.list(**list_kwargs)
+    while request is not None:
+        response = request.execute()
+        for item in response.get(key, []):
+            yield item
+        request = resource.list_next(previous_request=request,
+                                     previous_response=response)
```

Comment:
The complete content of this function should be enclosed in a `try-except` block, so that any zone discovery failure, does not crash the complete project discovery loop. Use this format for logging the exception:

```
            _log.error('Failed to fetch resource list; key: %s; '
                       'list_kwargs: %s; key_file_path: %s; '
                       'error: %s: %s', key, list_kwargs,
                       self._key_file_path, type(e).__name__, e)
```